pkg/backup/backup_compaction.go

```diff
@@ -0,0 +1,693 @@
+	encryptionOptions *kvpb.FileEncryptionOptions,
+	details jobspb.BackupDetails,
+) (mergedSST, error) {
+	var dirs []cloud.ExternalStorage
+	storeFiles := make([]storageccl.StoreFile, 0, len(entry.Files))
+	for idx := 0; idx < len(entry.Files); idx++ {
+		file := entry.Files[idx]
+		dir, err := execCtx.ExecCfg().DistSQLSrv.ExternalStorage(ctx, file.Dir)
+		if err != nil {
+			return mergedSST{}, err
+		}
+		dirs = append(dirs, dir)
+		storeFiles = append(storeFiles, storageccl.StoreFile{Store: dir, FilePath: file.Path})
+	}
+	iterOpts := storage.IterOptions{
+		// TODO (kev-cao): Come back and update this to range keys when
+		// SSTSinkKeyWriter has been updated to support range keys.
+		KeyTypes:   storage.IterKeyTypePointsOnly,
+		LowerBound: keys.LocalMax,
+		UpperBound: keys.MaxKey,
+	}
+	iter, err := storageccl.ExternalSSTReader(ctx, storeFiles, encryptionOptions, iterOpts)
+	if err != nil {
+		return mergedSST{}, err
+	}
+	compactionIter, err := storage.NewBackupCompactionIterator(iter, details.EndTime)
+	if err != nil {
+		return mergedSST{}, err
+	}
+	return mergedSST{
+		entry: entry,
+		iter:  compactionIter,
+		cleanup: func() {
+			log.VInfof(ctx, 1, "finished with and closing %d files in span %d %v", len(entry.Files), entry.ProgressIdx, entry.Span.String())
+			compactionIter.Close()
+			for _, dir := range dirs {
+				if err := dir.Close(); err != nil {
+					log.Warningf(ctx, "close export storage failed: %v", err)
+				}
+			}
+		},
+		completeUpTo: details.EndTime,
+	}, nil
+}
+
+// makeCompactionBackupDetails takes a chain of backups that are to be
+// compacted and returns a corresponding BackupDetails for the compacted
+// backup. It also takes in the job details for the last backup in its chain.
+func makeCompactionBackupDetails(
+	ctx context.Context,
+	lastIncDetails jobspb.BackupDetails,
+	manifests []backuppb.BackupManifest,
+	dest backupdest.ResolvedDestination,
+	encryptionOptions *jobspb.BackupEncryptionOptions,
+	kmsEnv cloud.KMSEnv,
+) (jobspb.BackupDetails, error) {
+	if len(manifests) == 0 {
+		return jobspb.BackupDetails{}, errors.New("no backup manifests to compact")
+	}
+	compactedDetails, err := updateBackupDetails(
+		ctx, lastIncDetails,
+		dest.CollectionURI,
+		dest.DefaultURI,
+		dest.ChosenSubdir,
+		dest.URIsByLocalityKV,
+		manifests,
+		encryptionOptions,
+		kmsEnv,
+	)
+	if err != nil {
+		return jobspb.BackupDetails{}, err
+	}
+	// The manifest returned by updateBackupDetails have its start and end times
+	// set to append to the chain. We need to update them to reflect the
+	// compacted backup's start and end times.
+	compactedDetails.StartTime = manifests[0].StartTime
+	compactedDetails.EndTime = manifests[len(manifests)-1].EndTime
+	return compactedDetails, nil
+}
+
+// compactIntroducedSpans takes a compacted backup manifest and the full chain of backups it belongs to and
+// computes the introduced spans for the compacted backup.
+func compactIntroducedSpans(
```

Comment:
We could add unit tests here, although it's covered in the `TestBackupCompaction` through the use of creating tables/indexes during incrementals...